extensions/dialog/common/client.mjs

```diff
@@ -15,42 +15,113 @@ class DialogModal extends Modal {
     const p = new Promise((resolve, reject) => {
-      btnOK.addEventListener('click', () => {
-        this.hide();
-        resolve(true);
-        return true;
-      });
+      for (const btn of btnList) {
+        const f = (btn) => {
+          // Extra layer of closure so we can pass btn in.
+          return () => {
+            if (!triggered) {
+              triggered = true;
+              this.hide();
+              resolve(btn);
+            }
+            return true;
+          };
+        };
+        btn.addEventListener('click', f(btn) );
+      }
     });
 
     this.show();
 
     return await p;
   }
 
+  /**
+   * Call to set the dialog to display a single choice dialog.
+   */
+  async displayAsSingleChoice(subject, message, buttonText) {
+    const result = await this.displayDialogInternal(() => {
+      this.subjectDom.innerHTML = subject;
+      this.msgDom.innerHTML = message;
+      this.btnContDom.innerHTML = `<button id="dialog-btn-OK">${buttonText}</button>`;
+      const btnOK = document.getElementById('dialog-btn-OK');
+
+      // Set the visibilities right.
+      this.btnContDom.classList.add('visible');
+
+      return [btnOK];
+    });
+
+    if (result === false) return {cancelled: true};
+    return {ok: true};
+  }
+
+  /**
+   * Call to set the dialog to display a multi choice dialog.
+   */
+  async displayAsMultiChoice(subject, message, choices) {
+    const result = await this.displayDialogInternal(() => {
+      this.subjectDom.innerHTML = subject;
+
+      let choicesHTML = '<ul>';
+      for (const {token, display} of choices) {
+        choicesHTML += `<li data-token=${token} id="dialog-btn-${token}" `;
+        choicesHTML += `class="dialog-choice-entry">${display}</li>`;
+      }
+      choicesHTML += '</ul>';
+      //this.choicesDom.innerHTML = choicesHTML;
+      this.msgDom.innerHTML = message + '<br />' + choicesHTML;
+
+      // Set the visibilities right.
+      this.btnContDom.classList.remove('visible');
+
+      const btnList = [];
+      for (const {token, display} of choices) {
+        btnList.push(document.getElementById(`dialog-btn-${token}`));
+      }
+
+      return btnList;
+    });
+
+    if (result === false) return {cancelled: true};
+    return {token: result.dataset.token};
+  }
+
+  /**
+   * Call to set the dialog to display an single line open-ended prompt.
+   */
+  async displayAsPrompt(subject, message, buttonText) {
+    const result = await this.displayDialogInternal(() => {
+      this.subjectDom.innerHTML = subject;
+      const promptHtml = '<br /><input type="text" class="dialog-textinput"' +
+        ' id="dialog-textinput" />';
+      this.msgDom.innerHTML = message + promptHtml;
+      this.btnContDom.innerHTML = `<button id="dialog-btn-OK">${buttonText}</button>`;
+      const btnOK = document.getElementById('dialog-btn-OK');
+
+      // Set the visibilities right.
+      this.btnContDom.classList.add('visible');
+
+      return [btnOK];
```

Comment:
Why return an `Array`?

Reply:
As mentioned, it's a callback that expects array for return.